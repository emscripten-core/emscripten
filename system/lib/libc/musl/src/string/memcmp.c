#if __EMSCRIPTEN__
#include <stdint.h>
#endif
#include <string.h>

int memcmp(const void *vl, const void *vr, size_t n)
{
	const unsigned char *l=vl, *r=vr;

// XXX EMSCRIPTEN: add an optimized version.
#if !defined(EMSCRIPTEN_OPTIMIZE_FOR_OZ) && !__has_feature(address_sanitizer)
	// If we have enough bytes, and everything is aligned, loop on words instead
	// of single bytes.
	if (n >= 4 && !((((uintptr_t)l) & 3) | (((uintptr_t)r) & 3))) {
		while (n >= 4) {
			if (*((uint32_t *)l) != *((uint32_t *)r)) {
				// Go to the single-byte loop to find the specific byte.
				break;
			}
			l += 4;
			r += 4;
			n -= 4;
		}
	}
#endif

#if defined(EMSCRIPTEN_OPTIMIZE_FOR_OZ)
#pragma clang loop unroll(disable)
#endif
	for (; n && *l == *r; n--, l++, r++);
	return n ? *l-*r : 0;
}
