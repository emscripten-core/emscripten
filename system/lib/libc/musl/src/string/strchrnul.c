#include <string.h>
#include <stdint.h>
#include <limits.h>
#include "libc.h"

#define ALIGN (sizeof(size_t))
#define ONES ((size_t)-1/UCHAR_MAX)
#define HIGHS (ONES * (UCHAR_MAX/2+1))
#define HASZERO(x) ((x)-ONES & ~(x) & HIGHS)

char *__strchrnul(const char *s, int c)
{
	c = (unsigned char)c;
	if (!c) return (char *)s + strlen(s);

/* XXX EMSCRIPTEN: add __has_feature check */
#if defined(__GNUC__) && !__has_feature(address_sanitizer)
	typedef size_t __attribute__((__may_alias__)) word;
	const word *w;
	for (; (uintptr_t)s % ALIGN; s++)
		if (!*s || *(unsigned char *)s == c) return (char *)s;
	size_t k = ONES * c;
	for (w = (void *)s; !HASZERO(*w) && !HASZERO(*w^k); w++);
	s = (void *)w;
#endif
	for (; *s && *(unsigned char *)s != c; s++);
	return (char *)s;
}

weak_alias(__strchrnul, strchrnul);
