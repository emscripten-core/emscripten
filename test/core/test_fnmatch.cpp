/*
 * Copyright 2014 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

// Begin test_fnmatch.cpp
#include <fnmatch.h>
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class TestCase {
public:
    TestCase(const string& pattern, const string& testString, int flags, int expected) :
        pattern(pattern),
        testString(testString),
        flags(flags),
        expected(expected)
    {}
    string pattern;
    string testString;
    int flags;
    int expected;
};

int main()
{
    vector<TestCase> testCases;

    testCases.push_back(TestCase("*","anything",0,0));
    testCases.push_back(TestCase("*.txt","readme.txt",0,0));
    testCases.push_back(TestCase("*.txt","readme.info",0,FNM_NOMATCH));
    testCases.push_back(TestCase("*.t?t","readme.txt",0,0));
    testCases.push_back(TestCase("*.t?t","readme.tot",0,0));
    testCases.push_back(TestCase("*.t?t","readme.txxt",0,FNM_NOMATCH));
    testCases.push_back(TestCase("[a-g]1","c1",0,0));
    testCases.push_back(TestCase("[a-g]1","i1",0,FNM_NOMATCH));
    testCases.push_back(TestCase("[!a-g]1","i1",0,0));
    testCases.push_back(TestCase("a\\*","anything",0,FNM_NOMATCH));
    testCases.push_back(TestCase("a\\*","a*",0,0));
    testCases.push_back(TestCase("a\\*","a*",FNM_NOESCAPE,FNM_NOMATCH));
    testCases.push_back(TestCase("a\\*","a\\*",FNM_NOESCAPE,0));
    testCases.push_back(TestCase("*readme","/etc/readme",0,0));
    testCases.push_back(TestCase("*readme","/etc/readme",FNM_PATHNAME,FNM_NOMATCH));
    testCases.push_back(TestCase("/*/readme","/etc/readme",FNM_PATHNAME,0));
    testCases.push_back(TestCase("*readme","/etc/.readme",0,0));
    testCases.push_back(TestCase("*readme",".readme",FNM_PERIOD,FNM_NOMATCH));
    testCases.push_back(TestCase("*.readme","/etc/.readme",FNM_PERIOD,0));
    testCases.push_back(TestCase("*.readme","/etc/.readme",FNM_PERIOD|FNM_PATHNAME,FNM_NOMATCH));
    testCases.push_back(TestCase("/*/.readme","/etc/.readme",FNM_PERIOD|FNM_PATHNAME,0));
    testCases.push_back(TestCase("ReAdME","readme",0,FNM_NOMATCH));

    bool pass = true;

    for (vector<TestCase>::const_iterator it = testCases.begin(); it != testCases.end(); ++it)
    {
        int result = fnmatch(it->pattern.c_str(), it->testString.c_str(), it->flags);
        if (result == it->expected)
            cout << "Pass: ";
        else
        {
            cout << "Fail: ";
            pass = false;
        }

        cout << "fnmatch(" << it->pattern << ", " << it->testString << ", "
             << it->flags << ") returned " << result << ", expected "
             << it->expected << endl;
    }

    if (pass)
    {
        cout << "All tests passed." << endl;
        return 0;
    }
    else
    {
        cout << "Some tests failed." << endl;
        return 1;
    }
}

