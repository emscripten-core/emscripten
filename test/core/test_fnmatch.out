Pass: fnmatch(*, anything, 0) returned 0, expected 0
Pass: fnmatch(*.txt, readme.txt, 0) returned 0, expected 0
Pass: fnmatch(*.txt, readme.info, 0) returned 1, expected 1
Pass: fnmatch(*.t?t, readme.txt, 0) returned 0, expected 0
Pass: fnmatch(*.t?t, readme.tot, 0) returned 0, expected 0
Pass: fnmatch(*.t?t, readme.txxt, 0) returned 1, expected 1
Pass: fnmatch([a-g]1, c1, 0) returned 0, expected 0
Pass: fnmatch([a-g]1, i1, 0) returned 1, expected 1
Pass: fnmatch([!a-g]1, i1, 0) returned 0, expected 0
Pass: fnmatch(a\*, anything, 0) returned 1, expected 1
Pass: fnmatch(a\*, a*, 0) returned 0, expected 0
Pass: fnmatch(a\*, a*, 2) returned 1, expected 1
Pass: fnmatch(a\*, a\*, 2) returned 0, expected 0
Pass: fnmatch(*readme, /etc/readme, 0) returned 0, expected 0
Pass: fnmatch(*readme, /etc/readme, 1) returned 1, expected 1
Pass: fnmatch(/*/readme, /etc/readme, 1) returned 0, expected 0
Pass: fnmatch(*readme, /etc/.readme, 0) returned 0, expected 0
Pass: fnmatch(*readme, .readme, 4) returned 1, expected 1
Pass: fnmatch(*.readme, /etc/.readme, 4) returned 0, expected 0
Pass: fnmatch(*.readme, /etc/.readme, 5) returned 1, expected 1
Pass: fnmatch(/*/.readme, /etc/.readme, 5) returned 0, expected 0
Pass: fnmatch(ReAdME, readme, 0) returned 1, expected 1
All tests passed.
